Framework/Algorithms/src/Q1DWeighted.cpp

```diff
@@ -154,199 +139,228 @@ void Q1DWeighted::exec() {
-          if (errorWeighting) {
-            // When using the error as weight we have:
-            //    w_i = 1/s_i^2   where s_i is the uncertainty on the ith pixel.
-            //
-            //    I(q_i) = (sum over i of I_i * w_i) / (sum over i of w_i)
-            //       where all pixels i contribute to the q_i bin, and I_i is
-            //       the intensity in the ith pixel.
-            //
-            //    delta I(q_i) = 1/sqrt( (sum over i of w_i) )  using simple
-            //    error propagation.
-            double err = 1.0;
-            if (EIn[j] > 0)
-              err = EIn[j];
-            w = 1.0 / (nSubPixels * nSubPixels * err * err);
-          }
+      // Each pixel might be sub-divided in the number of pixels given as input
+      // parameter (NPixelDivision x NPixelDivision)
+      for (int isub = 0; isub < nSubPixels * nSubPixels; ++isub) {
 
-          PARALLEL_CRITICAL(iqnorm) /* Write to shared memory - must protect */
-          {
-            lambdaIq[iq] += YIn[j] * w;
-            lambdaIqErr[iq] += w * w * EIn[j] * EIn[j];
-            XNorm[iq] += w;
+        // Find the position offset for this sub-pixel in real space
+        const double subY = pixelSizeY *
+                            ((isub % nSubPixels) - (nSubPixels - 1.0) / 2.0) /
+                            nSubPixels;
+        const double subX = pixelSizeX *
+                            (floor(static_cast<double>(isub) / nSubPixels) -
+                             (nSubPixels - 1.0) * 0.5) /
+                            nSubPixels;
+
+        // calculate Q
+        const V3D position = pos - V3D(subX, subY, 0.0) + correction;
+        const double sinTheta = sin(0.5 * position.angle(beamLine));
+        const double q = 4.0 * M_PI * sinTheta / wavelength;
+
+        if (q < qBinEdges.front() || q > qBinEdges.back()) {
+          continue;
+        }
 
-            // Fill in the wedge data
-            for (int iWedge = 0; iWedge < nWedges; iWedge++) {
-              double centerAngle = M_PI / nWedges * iWedge;
+        // after check above, no need to wrap this in try catch
+        const size_t k = VectorHelper::indexOfValueFromEdges(qBinEdges, q);
+
+        double w = 1.0;
+        if (errorWeighting) {
+          // When using the error as weight we have:
+          //    w_i = 1/s_i^2   where s_i is the uncertainty on the ith
+          //    pixel.
+          //
+          //    I(q_i) = (sum over i of I_i * w_i) / (sum over i of w_i)
+          //       where all pixels i contribute to the q_i bin, and I_i is
+          //       the intensity in the ith pixel.
+          //
+          //    delta I(q_i) = 1/sqrt( (sum over i of w_i) )  using simple
+          //    error propagation.
+          double err = 1.0;
+          if (EIn[j] > 0)
+            err = EIn[j];
+          w /= nSubPixels * nSubPixels * err * err;
+        }
+        PARALLEL_CRITICAL(iqnorm) /* Write to shared memory - must protect */
```

Comment:
It might be worth checking if breaking the critical section into two parts gives any performance benefits. The first block should protect the three lines below doing the actual writing while the second one inside the `if (nWedges != 0)` block protects the three writing lines there.